feat(message-input): dt-1092 message input vue 3

sanjeevdialpad opened the pull request:
# Message Input Recipe

<!--- Feel free to remove any unused sections -->

## :hammer_and_wrench: Type Of Change

<!--- Tick or place an `x` in all of the checkboxes that apply -->

- [ ] Fix
- [x] Feature
- [ ] Refactoring
- [ ] Documentation

## :book: Description

Message Input Recipe

## :bulb: Context

<!--- Describe the purpose of the changes -->
<!--- Why did we make these changes? -->
<!--- What problem(s) do they solve? -->

## :pencil: Checklist

<!--- Tick or place an `x` in all of the checkboxes that apply -->
<!--- Remove checkboxes that do not apply -->

- [ ] I have reviewed my changes
- [ ] I have added tests
- [ ] I have added all relevant documentation
- [ ] I have validated components with a screen reader
- [ ] I have validated components keyboard navigation
- [ ] I have considered the performance impact of my change
- [ ] I have checked that my change did not significantly increase bundle size
- [ ] I am exporting any new components or constants in the index.js in the component directory
- [ ] I am exporting any new components or constants in the index.js in the root

## :crystal_ball: Next Steps

<!--- Describe any future changes that need to be made after merging the PR -->

## :camera: Screenshots / GIFs

<!--- Mandatory for any UI work -->
<!--- Link any screenshots / GIFs below -->

## :link: Sources

<!--- Add any links to external reference material -->

sanjeevdialpad committed:
feat: message input

sanjeevdialpad commented:
This is the first pass or a more basic version of message input.

Still to be implemented as separate PR:
1. Character limit
2. Error banners

sanjeevdialpad committed:
Merge branch 'staging-vue3' into DT-1092-vue3

iropolo reviewed (dismissed):
Nice job here!, looks awesome, I will point a few visually catches before go through the code but it is great initial!


Emoji picker------/////


1. Emoji picker has some margins
This is added by the class `d-p16` in the `dt-popover-content` element. Im not sure if it should be there because it interfered the emoji-picker style.

<img width="547" alt="Screenshot 2023-06-12 at 11 31 55" src="https://github.com/dialpad/dialtone-vue/assets/89984179/9c1836d9-b1ba-4f9b-bd8d-33bc30767c28">

3. Emoji picker doesn't update the skin tone color.

4. When I select an emoji it doesnt appear correctly in the input field

<img width="561" alt="image" src="https://github.com/dialpad/dialtone-vue/assets/89984179/37714cf0-867a-4273-b023-ce4eb1ed2ddd">

5. Emoji picker input search should be focussed on mounted, this is already in place in the component but I don't see it working here, hmm... maybe I should expose a focus input search method so you can trigger it after open the emoji picker... but take a look if theres is something you can do on your side because the component already has a focus search input on mounted.

6. If I press enter on empty input search it throw an error:

<img width="561" alt="image" src="https://github.com/dialpad/dialtone-vue/assets/89984179/f87d7cef-f0c2-44bc-a3b2-2ee34a32ed88">


7. If I press enter in a tabset it close the emoji picker (it is not working as how the component should)

<img width="392" alt="image" src="https://github.com/dialpad/dialtone-vue/assets/89984179/3e32b98f-8dae-4656-9065-455336b733c2">

8. When I reach the last tabset in the emoji selector it doesn't jump to the skin selector, its jumping directly to the tabs (it is not working as how the component should)

9. The emoji picker has a color issue on the fixed label side 

<img width="392" alt="Screenshot 2023-06-12 at 11 42 40" src="https://github.com/dialpad/dialtone-vue/assets/89984179/e3e82385-a372-4b56-b18c-6019206f67b2">

```
.d-popover__dialog {
    --popover-color-background: var(--bgc-secondary);
```

Seems like is the popover which is adding this things, I guess you can place a white background color to override these in the `dt-popover-content` element.

10. The input search persist the previous search, it should be empty.

11. The emoji picker trough an error if you press enter in a no result emoji list.

<img width="392" alt="image" src="https://github.com/dialpad/dialtone-vue/assets/89984179/fad16918-8877-4de5-92c8-c99dc9106905">

<img width="446" alt="image" src="https://github.com/dialpad/dialtone-vue/assets/89984179/9c0e9798-da39-4ec7-b4ea-f0c0ba5ec249">

sanjeevdialpad commented:
@iropolo thank you for letting me know of these errors. this is very helpful.

The emoji picker specific errors excluding padding are related to the component which isnt part of this component/Recipe. I wonder if i should open bugs for it separately. @braddialpad 

I will go through each of the errors reported by you and create tickets or solve it if possible in this PR itself.

iropolo commented:
> @iropolo thank you for letting me know of these errors. this is very helpful.
> 
> The emoji picker specific errors excluding padding are related to the component which isnt part of this component/Recipe. I wonder if i should open bugs for it separately. @braddialpad
> 
> I will go through each of the errors reported by you and create tickets or solve it if possible in this PR itself.

Lets check this in parts....

3 )
If you add this in the `<dt-emoji-picker`
```
:skin-tone="skinTone"
@skin-tone="skinTone = $event"
```

and this inside `data () {`

`skinTone: 'Default',`

You fixed the skin tone change issue...

5-7-8)

Lets see now regarding the keyboard accesibility.
This is being affected by tippy somehow, because if you extract the `<dt-emoji-picker` from the `<template #content>` and put it outside, somewhere with a `v-if="emojiPickerOpened"` you will see the component mounts as expect and without problems, the input search is focused and the navigation with keyboard is working.


6-11)
Regarding the error on empty or `No result` emoji after press Enter it is because this method is not validating a valid emoji:

```
onSelectEmoji (emoji) {
      this.inputValue = this.inputValue + emoji.shortname;
      this.emojiPickerOpened = false;
    },
```

It can be simply fixed adding this `if (!emoji) { return; }` or sth..

sanjeevdialpad commented:
@iropolo ty. The accessibility issue i was still figuring out. Your solution fixes it.

On a unrelated note, all the labels currently are hardcoded on this recipe, should we pass it down from props since we need product to do i18n ?

iropolo commented:
> 

Certainly yes, this should not be hardcoded in the final product.
All labels should be managed by props to give the possibility to i18n to write it.

sanjeevdialpad committed:
fix: emoji picker issues

sanjeevdialpad committed:
feat: character limit feature in recipe

sanjeevdialpad commented:
https://dialpad.atlassian.net/browse/DIALTONE-1094


Character Limit code has been added for the recipe.

sanjeevdialpad committed:
Revert "feat: character limit feature in recipe"

This reverts commit 757338eebe80471d8633fa6ea426b27c3f28f6aa.

sanjeevdialpad commented:
Creating separate PRs for characterLimit and error message.

Character Limit logic: https://github.com/dialpad/dialtone-vue/pull/1019

juliodialpad committed:
docs: fix broken story

semantic-release-bot committed:
chore(release): 3.71.0

# [3.71.0](https://github.com/dialpad/dialtone-vue/compare/v3.70.0...v3.71.0) (2023-06-13)

### Bug Fixes

* add placeholder text for richtexteditor - vue3 ([#1005](https://github.com/dialpad/dialtone-vue/issues/1005)) ([ae5d586](https://github.com/dialpad/dialtone-vue/commit/ae5d586e2ec368fc283d4fb609769bb8c62c3037))
* **Contact Row, Group Row, General Row:** null on description - vue3 ([#1014](https://github.com/dialpad/dialtone-vue/issues/1014)) ([ac6d084](https://github.com/dialpad/dialtone-vue/commit/ac6d08462184f2324b033c129a2c853264125044))
* **Modal:** secondary action button kind ([#1012](https://github.com/dialpad/dialtone-vue/issues/1012)) ([0f51007](https://github.com/dialpad/dialtone-vue/commit/0f510075861bf3b61c0e67117c2fc914f5ef58a0))
* **Modal:** show event and initial-focus-element - vue3 ([#1010](https://github.com/dialpad/dialtone-vue/issues/1010)) ([00a4553](https://github.com/dialpad/dialtone-vue/commit/00a4553e498f0cbebc499b7a1d299f5831fa0ff3))

### Documentation

* fix broken story ([318974c](https://github.com/dialpad/dialtone-vue/commit/318974cfe7b8a757f319bb8f78e2b473186d8fda))

### Features

* **Feed Item Row:** state changes - vue3 ([#1006](https://github.com/dialpad/dialtone-vue/issues/1006)) ([13e4edf](https://github.com/dialpad/dialtone-vue/commit/13e4edf304f250140db1208b2be84fe2b3616ee8))
* **Rich Text Editor:** link extension vue3 ([#1008](https://github.com/dialpad/dialtone-vue/issues/1008)) ([b4c1397](https://github.com/dialpad/dialtone-vue/commit/b4c13970db9ad6e88bb8ce316c50ee430e46d872))

juliodialpad committed:
refactor(list-item): remove utility classes (#1018)

iropolo commented:
Created #1020 to unblock this PR and continue working.

sanjeevdialpad committed:
feat(DT-1094): handle character limit logic in Message Input Recipe (#1019)

braddialpad commented on recipes/conversation_view/message_input/message_input.mdx:
```diff
@@ -0,0 +1,30 @@
+# Message Input Recipe
+
+<Subtitle>
+  Message Input recipe wraps the Rich Text Editor to be used for conversations.
```
In theory doesn't have to be used only for conversations. ex: writing notes or something

iropolo committed:
feat(emoji-picker): with popover variant and code improvement (#1020)

* feat(emoji-picker): with popover variant and code improvement

* feat(emoji-picker): clean code and documentation update

* feat(emoji-picker): doc lint fix

* feat(emoji-picker): code snippet in doc

---------

Co-authored-by: iropolo <ignacio.ropolo@dialpad.com>

braddialpad commented on recipes/conversation_view/message_input/message_input.vue:
```diff
@@ -0,0 +1,359 @@
+
+    // Emoji picker props
+
+    /**
+     * tab labels for emoji
+     */
+    tabSetLabels: {
+      type: Array,
+      default: () => [
+        'Most recently used',
+        'Smileys and people',
+        'Nature',
+        'Food',
+        'Activity',
+        'Travel',
+        'Objects',
+        'Symbols',
+        'Flags',
+      ],
+    },
+
+    skinSelectorButtonTooltipLabel: {
+      type: String,
+      default: 'Change default skin tone',
+    },
+
+    searchNoResultsLabel: {
+      type: String,
+      default: 'No results',
+    },
+
+    searchResultsLabel: {
+      type: String,
+      default: 'Search results',
+    },
+
+    searchPlaceholderLabel: {
+      type: String,
+      default: 'Search...',
+    },
```
add emoji in front of the name of each of these, it is not clear in the docs that they are for the emoji picker portion of this component

ex: `searchPlaceholderLabel` = `emojiSearchPlaceholderLabel`

sanjeevdialpad replied:
changed.

braddialpad commented on recipes/conversation_view/message_input/message_input_default.story.vue:
```diff
@@ -0,0 +1,52 @@
+        #footerLeft
+      >
+        <div class="message-input-recipe-subtitle d-fc-black-500">
+          <b>Dwight</b> is typing...
```
lol I love how you always use dwight for examples.

sanjeevdialpad replied:
Not just notes, wait till you see DPM beta datastore. 😛

sanjeevdialpad committed:
disable fix

sanjeevdialpad committed:
Merge remote-tracking branch 'origin/staging-vue3' into DT-1092-vue3

sanjeevdialpad committed:
aria-disabled

braddialpad reviewed (commented):
Just adding some doc stuff since @iropolo has already reviewed thoroughly.

Looks great!

sanjeevdialpad committed:
emoji picker changes

sanjeevdialpad commented:
@iropolo Emoji pickers been fixed.

sanjeevdialpad committed:
focus events

braddialpad reviewed (approved):
Looks good by me, you just need to fix the accessibility tests.

braddialpad commented on recipes/conversation_view/message_input/message_input.vue:
```diff
@@ -0,0 +1,379 @@
+      <section class="d-d-flex d-jc-space-between d-mx8 d-my4">
+        <!-- Left content -->
+        <div class="d-d-flex">
+          <dt-button
+            size="sm"
+            circle
+            importance="clear"
+          >
```
These buttons need aria-label, will have to be a prop because of i18n

sanjeevdialpad replied:
got it.

braddialpad commented:
Just noticed there is also an error occurring on this page.

```
button.vue:314 You cannot have a circle button with kind: default and importance: primary as it does not exist in our design system.
```

sanjeevdialpad committed:
accessibility test fix

sanjeevdialpad committed:
fix button circle design error

sanjeevdialpad committed:
aria labelled for input field

sanjeevdialpad committed:
Merge branch 'staging-vue3' into DT-1092-vue3

sanjeevdialpad commented:
Will be making the error banner and the image selector in a different PR. fixed all issues.
